superset/sql_parse.py

```diff
@@ -284,19 +286,64 @@ def _extract_tables_from_sql(self) -> set[Table]:
         Extract all table references in a query.
 
         Note: this uses sqlglot, since it's better at catching more edge cases.
+
+        Due to Jinja templating a multiphase approach is necessary as the referenced SQL
+        statement is likely ill-defined (due to the presence of the Jinja macros) and
+        thus non-parsable by SQLGlot.
+
+        Firstly, we extract any tables referenced within the confines of specific Jinja
+        macros. Secondly, we replace these non-SQL Jinja calls with a pseudo-benign SQL
+        expression to help ensure that the resulting SQL statements are parsable by
+        SQLGlot.
         """
+
+        from superset.jinja_context import (  # pylint: disable=import-outside-toplevel
+            get_template_processor,
+        )
+
+        tables = set()
+        sql = self.stripped()
+
+        # Mock the required database as the processor signature is exposed publically.
+        processor = get_template_processor(database=Mock(backend=self._dialect))
+        template = processor.env.parse(sql)
+
+        for node in template.find_all(nodes.Call):
+            if isinstance(node.node, nodes.Getattr) and node.node.attr in (
+                "latest_partition",
+                "latest_sub_partition",
+            ):
+                # Extract the table referenced in the macro.
+                tables.add(
+                    Table(
+                        *[
+                            remove_quotes(part)
+                            for part in node.args[0].value.split(".")[::-1]
+                            if len(node.args) == 1
+                        ]
+                    )
+                )
+
+                # Replace the potentially problematic Jinja macro with benign SQL.
+                node.__class__ = nodes.TemplateData
+                node.fields = nodes.TemplateData.fields
+                node.data = "NULL"
+
+        sql = processor.process_template(template)
+
         try:
-            statements = parse(self.stripped(), dialect=self._dialect)
+            tables.update(
+                [
+                    table
+                    for statement in parse(sql, dialect=self._dialect)
+                    for table in self._extract_tables_from_statement(statement)
+                    if statement
+                ]
+            )
         except SqlglotError:
-            logger.warning("Unable to parse SQL (%s): %s", self._dialect, self.sql)
-            return set()
+            logger.warning("Unable to parse SQL (%s): %s", self._dialect, sql)
```

Comment:
I think we should throw an exception if a sqlglot or Jinja parsing error occurs instead of returning an empty set. The main reason is because we can have a valid SQL query with no tables, which would also evaluate to an empty set. This is a distinct situation from the inability to determine the existence of tables, and the implications can vary. I think it's the responsibility of the caller to deal with both scenarios.

Reply:
@michael-s-molina this case is being handled in @betodealmeida's https://github.com/apache/superset/pull/26748.